dojo/problem/views.py

```diff
@@ -116,14 +140,34 @@ def get_problems(self, request: HttpRequest):
 class ListClosedProblems(ListProblems):
     filter_name = "Closed"
 
-    def get_problems(self, request: HttpRequest):
-        list_problem = []
+    def get_problems(self, request: HttpRequest, products=None):
+        finding_ids_in_problems = [
+            fid for p in self.problems_map.values() for fid in p.finding_ids
+        ]
+        if products:
+            user_findings = Finding.objects.filter(
+                id__in=finding_ids_in_problems,
+                test__engagement__product__in=products,
+                active=True,
+            )
+        else:
+            user_findings = Finding.objects.filter(
+                id__in=finding_ids_in_problems,
+                active=True,
+            )
+        allowed_finding_ids = set(user_findings.values_list("id", flat=True))
         active_findings = set(
-            Finding.objects.filter(id__in=[fid for p in self.problems_map.values() for fid in p.finding_ids], active=True)
-            .values_list("id", flat=True),
+            user_findings.filter(active=True).values_list("id", flat=True),
         )
+
+        list_problem = []
         for _, problem in self.problems_map.items():
-            if not any(finding_id in active_findings for finding_id in problem.finding_ids):
+            relevant_finding_ids = [
+                fid for fid in problem.finding_ids if fid in allowed_finding_ids
+            ]
+            if not relevant_finding_ids:
+                continue
+            if not any(fid in active_findings for fid in relevant_finding_ids):
```

Comment:
I find this code very confusing. There are list comprehensions everywhere. It seems worth trying to compute the sets once and call them something like `active_fids`, `user_fids`, `problem_fids`, and then set `intersection` operations to build resulting sets.